src/search.cpp

```diff
@@ -990,12 +990,27 @@ namespace {
           if (moveCount == 1 || value > alpha)
           {
               rm.score = value;
-              rm.pv.resize(1);
 
               assert((ss+1)->pv);
-
+              std::vector<Move> moves;
+              moves.push_back(rm.pv[0]);
               for (Move* m = (ss+1)->pv; *m != MOVE_NONE; ++m)
-                  rm.pv.push_back(*m);
+                  moves.push_back(*m);
+
+              // Try hard to maintain a long PV for analysis.  If the new PV is valid (alpha < v < beta),
+              // or longer, then always use it.  Otherwise, check that the moves in the new PV match
+              // the stored PV, and if they do, just keep the old full PV for display purposes.
+              rm.validPvLength = moves.size();
+              if ((value > alpha && value < beta) || moves.size() > rm.pv.size()) {
+                  rm.pv = moves;
+              } else {
+                  for (size_t i = 0; i < std::min(rm.pv.size(), moves.size()); ++i) {
```

Comment:
In the "else" block it is already guaranteed that moves.size() is not bigger than rm.pv.size() so it is possible to get rid of the std::min.